### PR TITLE
Deflake faster joins tests by leaving rooms as part of test cleanup

### PR DESCRIPTION
**Pull request by squahtx**
Many of the faster joins test flakes are due to the homeserver under
test failing to contact Complement homeservers after they have been
torn down. When this happens, subsequent tests can fail if they use a
Complement homeserver that happens to have the same hostname:port as one
which the homeserver under test has previously marked as offline.

Make the homeserver under test leave rooms at the end of each test, so
that it stops trying to contact the Complement homeservers in the room.

Signed-off-by: Sean Quah <seanq@matrix.org>

---

Reviewable commit by commit.

NB: you can tease out the flakes by forcing Complement to reuse ports whenever possible:
```diff
diff --git a/internal/federation/server.go b/internal/federation/server.go
index ed7974d..4c4ccbe 100644
--- a/internal/federation/server.go
+++ b/internal/federation/server.go
@@ -455,7 +455,10 @@ func (s *Server) Listen() (cancel func()) {
        var wg sync.WaitGroup
        wg.Add(1)

-       ln, err := net.Listen("tcp", ":0") //nolint
+       ln, err := net.Listen("tcp", ":63333") //nolint
+       if err != nil {
+               ln, err = net.Listen("tcp", ":63334") //nolint
+       }
        if err != nil {
                s.t.Fatalf("ListenFederationServer: net.Listen failed: %s", err)
        }
```

### CODE REVIEW

File: tests/federation_room_join_partial_state_test.go [hunk 1]
```diff
@@ -121,6 +121,44 @@ func (s *server) AddEDUHandler(eduHandler func(gomatrixserverlib.EDU) bool) func
...
 }
 
+// WithWaitForLeave runs the given action and waits for the user to leave the room.
+func (s *server) WithWaitForLeave(
+	t *testing.T, room *federation.ServerRoom, userID string, leaveAction func(),
+) {
+	leaveChannel := make(chan *gomatrixserverlib.Event, 10)
+	removePDUHandler := s.AddPDUHandler(
+		func(e *gomatrixserverlib.Event) bool {
+			if membership, _ := e.Membership(); e.Type() == "m.room.member" &&
```
Comment (DMRobertson) [inline tests/federation_room_join_partial_state_test.go:131]: what's the second thing returned by `Membership()` that gets bound(?) to `_` -- a  state key?

**Review by DMRobertson — APPROVED** [review]
LGTM, huge thanks for this!

File: tests/federation_room_join_partial_state_test.go [hunk 2]
```diff
@@ -52,17 +66,61 @@ func createTestServer(t *testing.T, deployment *docker.Deployment, opts ...func(
...
+// AddPDUHandler adds a PDU callback that returns `true` if it expected the given PDU.
+// When a PDU is received which is not expected by any PDU callback, the ongoing test is failed.
+// Returns a function to remove the PDU callback.
+func (s *server) AddPDUHandler(pduHandler func(*gomatrixserverlib.Event) bool) func() {
+	pduHandlerKey := s.nextPDUHandlerKey
+	s.nextPDUHandlerKey++
+	s.pduHandlers[pduHandlerKey] = pduHandler
+
+	return func() {
+		delete(s.pduHandlers, pduHandlerKey)
```
[the remark this answers is not in the document]
Comment (squahtx) [reply tests/federation_room_join_partial_state_test.go:107]: I imagine so

File: tests/federation_room_join_partial_state_test.go (hunk 1 above)
Comment (squahtx) [reply tests/federation_room_join_partial_state_test.go:131]: It's `err` (surprise!). `membership` is `""` when there's no such state entry, so I elided the `err` check.

**Comment by squahtx** [comment]
Thanks for the review!

**Pull request merged**